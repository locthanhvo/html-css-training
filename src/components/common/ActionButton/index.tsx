import { memo } from 'react';
import { Button, ButtonProps } from '@chakra-ui/react';

interface ActionButtonProps extends ButtonProps {
  type?: 'button' | 'submit' | 'reset';
  size?: 'sm' | 'md' | 'lg' | 'xl';
  isLoading?: boolean;
  isDisabled?: boolean;
  title?: string;
  bgColor?: string;
  color?: string;
  borderColor?: string;
  onClick?: () => void;
}

const ActionButton = ({
  title = 'Save',
  type = 'button',
  size = 'sm',
  isLoading = false,
  isDisabled = false,
  bgColor,
  color,
  borderColor,
  onClick,
  ...props
}: ActionButtonProps) => {
  return (
    <Button
      type={type}
      name={title}
      size={size}
      bgColor={bgColor}
      color={color}
      border="1px solid"
      borderColor={borderColor}
      onClick={onClick}
      isLoading={isLoading}
      isDisabled={isDisabled}
      {...props}
    >
      {title}
    </Button>
  );
};

export default memo(ActionButton);
