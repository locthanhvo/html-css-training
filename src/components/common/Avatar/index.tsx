import { Center, Image } from '@chakra-ui/react';
import { memo } from 'react';

interface AvatarProps {
  url?: string;
  w?: number;
  h?: number;
}

const Avatar = ({ url, w = 20, h = 20 }: AvatarProps) => {
  return (
    <Center w={w} h={h} borderRadius="50%">
      <Image borderRadius="50%" src={url} />
    </Center>
  );
};

export default memo(Avatar);
