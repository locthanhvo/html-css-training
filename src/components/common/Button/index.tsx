import { ReactNode } from 'react'
import './button.css'

interface Props {
  type?: 'button' | 'submit'
  children: ReactNode
  variants: 'primary' | 'secondary' | 'default' | 'danger'
  onClick: (e: React.FormEvent) => void
  isActive?: boolean
  isDisabled?: boolean
}

const Button = ({ children, type = 'button', variants, isActive, isDisabled, onClick }: Props) => {
  const activeClass = isActive ? 'filter-active' : ''
  const disabledClass = isDisabled ? 'btn-disabled' : ''

  return (
    <button
      disabled={isDisabled}
      type={type}
      className={`btn d-flex flex-center c-pointer btn-${variants} ${activeClass} ${disabledClass}`}
      onClick={onClick}
    >
      {children}
    </button>
  )
}

export default Button
