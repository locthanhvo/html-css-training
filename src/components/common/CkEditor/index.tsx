import { memo, useCallback } from 'react';
import { CKEditor } from '@ckeditor/ckeditor5-react';
import { FormControl, FormLabel } from '@chakra-ui/react';
import ClassicEditor from '@ckeditor/ckeditor5-build-classic';

interface CKEditorWrapperProps {
  onChange?: (content: string) => void;
  initialValue?: string;
  label?: string;
}

const CKEditorWrapper = ({
  onChange,
  label = '',
  initialValue = '',
}: CKEditorWrapperProps) => {
  const handleEditorChange = useCallback(
    (_event: unknown, editor: ClassicEditor) => {
      const data = editor.getData();
      onChange?.(data);
    },
    [onChange],
  );

  return (
    <FormControl maxW={430}>
      {label && (
        <FormLabel
          fontSize="base"
          fontWeight="regular"
          marginInlineEnd={0}
          minW="max-content"
        >
          {label}
        </FormLabel>
      )}
      <CKEditor
        editor={ClassicEditor}
        data={initialValue}
        onChange={handleEditorChange}
      />
    </FormControl>
  );
};

export default memo(CKEditorWrapper);
