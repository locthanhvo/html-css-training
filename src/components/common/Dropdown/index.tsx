import { ChangeEvent, memo, useState, useCallback } from 'react';
import {
  Checkbox,
  CheckboxGroup,
  Menu,
  MenuButton,
  MenuList,
  MenuItem,
  Button,
} from '@chakra-ui/react';
import { ISelectField, ITask } from '@/types';

interface DropdownProps {
  defaultValue?: { name: string; image?: string; value: string }[];
  options: { name: string; image?: string; value: string }[];
  name: string;
  onChange?: (selectField: ISelectField<ITask>) => void;
  field: keyof ITask;
}

const Dropdown = ({
  defaultValue = [],
  options,
  name,
  onChange,
  field,
}: DropdownProps) => {
  const [selectedOptions, setSelectedOptions] =
    useState<{ name: string; image?: string; value: string }[]>(defaultValue);

  const handleSelectOption = useCallback(
    (option: { name: string; image?: string; value: string }) => {
      setSelectedOptions((prevOptions) => {
        const newOptions = prevOptions.some(
          (item) => item.value === option.value,
        )
          ? prevOptions.filter((item) => item.value !== option.value)
          : [...prevOptions, option];
        onChange?.({
          field,
          data: newOptions,
          isError: false,
          onChange: () => {},
        });
        return newOptions;
      });
    },
    [onChange, field],
  );

  const isChecked = useCallback(
    (option: { name: string; image?: string; value: string }) =>
      selectedOptions.some((item) => item.value === option.value),
    [selectedOptions],
  );

  const handleCheckboxClick = (
    e: ChangeEvent<HTMLInputElement>,
    option: { name: string; image?: string; value: string },
  ) => {
    e.stopPropagation();
    handleSelectOption(option);
  };

  return (
    <Menu closeOnSelect={false}>
      <MenuButton
        as={Button}
        w={168}
        border="2px solid"
        borderColor="lightGray"
        color="primary"
        fontSize="md"
      >
        {name}
      </MenuButton>
      <MenuList>
        <CheckboxGroup value={selectedOptions.map((option) => option.value)}>
          {options.map((option) => (
            <MenuItem key={option.value}>
              <Checkbox
                isChecked={isChecked(option)}
                onChange={(e) => handleCheckboxClick(e, option)}
              >
                {option.name}
              </Checkbox>
            </MenuItem>
          ))}
        </CheckboxGroup>
      </MenuList>
    </Menu>
  );
};

export default memo(Dropdown);
