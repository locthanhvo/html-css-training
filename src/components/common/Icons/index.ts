export { default as UserProfileIcon } from './UserProfileIcon';
export { default as BuildingIcon } from './BuildingIcon';
export { default as WorldIcon } from './WorldIcon';
export { default as ForwardArrowIcon } from './ForwardArrowIcon';
export { default as PlusCircleIcon } from './PlusCircleIcon';
export { default as CompassIcon } from './CompassIcon';
export { default as FavoriteIcon } from './FavoriteIcon';
export { default as MenuIcon } from './MenuIcon';
export { default as MessageIcon } from './MessageIcon';
export { default as NotifyIcon } from './NotifyIcon';
export { default as SearchIcon } from './SearchIcon';
export { default as StarIcon } from './StarIcon';
export { default as StatisticsIcon } from './StatisticsIcon';
export { default as MeatballsMenuIcon } from './MeatballsMenuIcon';
export { default as BarChartIcon } from './BarChartIcon';
export { default as ChecklistIcon } from './ChecklistIcon';
export { default as CloudIcon } from './CloudIcon';
export { default as EmailIcon } from './EmailIcon';
export { default as UserGroupIcon } from './UserGroupIcon';
export { default as CalendarIcon } from './CalendarIcon';
export { default as SettingIcon } from './SettingIcon';
export { default as PhoneIcon } from './PhoneIcon';
export { default as BookIcon } from './BookIcon';
export { default as TelegramIcon } from './TelegramIcon';
export { default as WhatsappIcon } from './WhatsappIcon';
export { default as ClockIcon } from './ClockIcon';
export { default as AttachmentIcon } from './AttachmentIcon';
export { default as HamburgerMenuIcon } from './HamburgerMenuIcon';
export { default as ProfileIcon } from './ProfileIcon';
export { default as PlusIcon } from './PlusIcon';
export { default as BurgerMenuLeftIcon } from './BurgerMenuLeftIcon';
export { default as FlagIcon } from './FlagIcon';
export { default as CloseIcon } from './CloseIcon';
