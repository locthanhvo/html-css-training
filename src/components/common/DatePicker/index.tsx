import { useState, memo } from 'react';
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
  PopoverBody,
  Stack,
  Text,
  Center,
  Flex,
} from '@chakra-ui/react';
import { Controller, Control } from 'react-hook-form';

// Icons
import { ClockIcon } from '@/components/common/Icons';

// Utils
import { calculateDaysLeft } from '@/utils';

// Components
import InputField from '@/components/common/InputField';

// Types
import { IInputField, ITask } from '@/types';

interface DatePickerProps {
  control: Control<ITask>;
  variant?: 'icon' | 'button';
  onChange?: (inputField: IInputField<ITask>) => void;
}

const DatePicker = ({
  control,
  variant = 'icon',
  onChange,
}: DatePickerProps) => {
  const [startDate, setStartDate] = useState<string>('');
  const [endDate, setEndDate] = useState<string>('');

  return (
    <Popover>
      <PopoverTrigger>
        <Center
          gap={3}
          w={variant === 'button' ? 168 : 'none'}
          h={8}
          border={variant === 'button' ? '2px solid' : 'none'}
          borderColor="lightGray"
          borderRadius="md"
          cursor="pointer"
        >
          <Flex alignItems="center" justifyContent="center" gap={2}>
            {variant === 'icon' && <ClockIcon />}

            {variant === 'icon' && startDate && endDate && (
              <Text fontSize="sm" color="coolGray" textAlign="center">
                {calculateDaysLeft(startDate, endDate)}
              </Text>
            )}
          </Flex>
          {variant === 'button' && (
            <Text fontSize="base" color="primary" textAlign="center">
              Date
            </Text>
          )}
        </Center>
      </PopoverTrigger>
      <PopoverContent w={230}>
        <PopoverBody>
          <Stack spacing={4}>
            <Controller
              name="startDate"
              control={control}
              render={({ field }) => (
                <InputField
                  label="Start Date"
                  fontSize="base"
                  color="primary"
                  border="2px solid"
                  borderColor="primary"
                  {...field}
                  onChange={(data) => {
                    setStartDate(data);
                    onChange?.({
                      field: 'startDate',
                      data,
                      isError: false,
                      onChange: field.onChange,
                    });
                  }}
                  type="date"
                />
              )}
            />

            <Controller
              name="endDate"
              control={control}
              render={({ field }) => (
                <InputField
                  label="End Date"
                  fontSize="base"
                  color="primary"
                  border="2px solid"
                  borderColor="primary"
                  {...field}
                  onChange={(data) => {
                    setEndDate(data);
                    onChange?.({
                      field: 'endDate',
                      data,
                      isError: false,
                      onChange: field.onChange,
                    });
                  }}
                  type="date"
                  min={startDate}
                  isDisabled={!startDate}
                />
              )}
            />
          </Stack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(DatePicker);
