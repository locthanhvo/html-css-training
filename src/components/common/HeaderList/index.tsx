import { memo } from 'react';
import { Button, Flex, Text, VStack } from '@chakra-ui/react';

// Icons
import { PlusIcon, MeatballsMenuIcon } from '@/components/common/Icons';

interface HeaderListProps {
  title?: string;
  total?: number;
  color?: string;
  isDisabled?: boolean;
  onClick?: () => void;
}

const HeaderList = ({
  onClick,
  title = '',
  color = 'primary',
  total = 0,
  isDisabled = false,
}: HeaderListProps) => {
  return (
    <VStack
      position="relative"
      w={230}
      bgColor="white"
      flexDirection="row"
      justifyContent="space-between"
      alignItems="center"
      borderRadius="lg"
      boxShadow="md"
      py={4}
      px={3}
      pr={2}
      _before={{
        content: '""',
        position: 'absolute',
        left: 0,
        top: '50%',
        transform: 'translateY(-50%)',
        height: '50%',
        width: '3px',
        bg: color,
        borderRadius: 'lg',
      }}
    >
      <Flex gap={3} justifyContent="center" alignItems="center">
        <Text size="base" color="steelBlue" textTransform="uppercase">
          {title}
        </Text>

        <Flex
          w={30}
          h={30}
          alignItems="center"
          justifyContent="center"
          borderRadius="50%"
          boxShadow="md"
          bgColor="white"
          fontSize="sm"
          color="primary"
        >
          {total}
        </Flex>
      </Flex>

      <Flex>
        <Button name="menu" variant="outline">
          <MeatballsMenuIcon />
        </Button>

        <Button
          name="add-task"
          variant="outline"
          onClick={onClick}
          isDisabled={isDisabled}
        >
          <PlusIcon />
        </Button>
      </Flex>
    </VStack>
  );
};

export default memo(HeaderList);
