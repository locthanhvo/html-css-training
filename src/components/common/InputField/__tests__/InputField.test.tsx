import { fireEvent, render, RenderResult } from '@testing-library/react';

// Icons
import { ForwardArrowIcon, SearchIcon } from '@/components/common/Icons';

// Components
import InputField from '..';

const mockOnChange = jest.fn();
const props = {
  label: 'Search',
  name: 'search',
  placeholder: 'Search Tasks',
  onChange: mockOnChange,
};

describe('InputField component', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(<InputField {...props} />);
  });

  it('renders correctly with default props', () => {
    const { container } = renderResult;
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with left icon and right icon', () => {
    renderResult = render(
      <InputField
        {...props}
        leftIcon={<SearchIcon />}
        rightIcon={<ForwardArrowIcon />}
      />,
    );
    const { container } = renderResult;
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with error message', () => {
    renderResult = render(
      <InputField
        {...props}
        isError
        isValidate
        errorMessages="Field is required"
      />,
    );
    const { container } = renderResult;
    expect(container).toMatchSnapshot();
  });

  it('should call onChange when input value is changed', async () => {
    const { getByLabelText } = renderResult;
    fireEvent.change(getByLabelText('Search'), {
      target: { value: 'New task' },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });
});
