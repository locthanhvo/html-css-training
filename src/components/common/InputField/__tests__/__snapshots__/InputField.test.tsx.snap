// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField component renders correctly with default props 1`] = `
<div>
  <div
    class="chakra-form-control css-17qx7uq"
    role="group"
  >
    <label
      class="chakra-form__label css-zecyy3"
      for="field-:r0:"
      id="field-:r0:-label"
    >
      Search
    </label>
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <input
        class="chakra-input css-1m48l39"
        id="field-:r0:"
        name="search"
        placeholder="Search Tasks"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`InputField component renders correctly with error message 1`] = `
<div>
  <div
    class="chakra-form-control css-17qx7uq"
    data-invalid=""
    role="group"
  >
    <label
      class="chakra-form__label css-zecyy3"
      data-invalid=""
      for="field-:r4:"
      id="field-:r4:-label"
    >
      Search
    </label>
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <input
        aria-describedby="field-:r4:-feedback"
        aria-invalid="true"
        class="chakra-input css-1m48l39"
        id="field-:r4:"
        name="search"
        placeholder="Search Tasks"
        type="text"
      />
    </div>
    <div
      aria-live="polite"
      class="chakra-form__error-message css-1qc3jta"
      id="field-:r4:-feedback"
    >
      Field is required
    </div>
  </div>
</div>
`;

exports[`InputField component renders correctly with left icon and right icon 1`] = `
<div>
  <div
    class="chakra-form-control css-17qx7uq"
    role="group"
  >
    <label
      class="chakra-form__label css-zecyy3"
      for="field-:r2:"
      id="field-:r2:-label"
    >
      Search
    </label>
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <div
        class="chakra-input__left-element css-1yuzhmq"
      >
        <svg
          fill="none"
          height="31"
          viewBox="0 0 30 31"
          width="30"
        >
          <path
            clip-rule="evenodd"
            d="M18.5006 17.5006H17.7106L17.4306 17.2306C18.6306 15.8306 19.2506 13.9206 18.9106 11.8906C18.4406 9.11063 16.1206 6.89063 13.3206 6.55063C9.09063 6.03063 5.53063 9.59063 6.05063 13.8206C6.39063 16.6206 8.61063 18.9406 11.3906 19.4106C13.4206 19.7506 15.3306 19.1306 16.7306 17.9306L17.0006 18.2106V19.0006L21.2506 23.2506C21.6606 23.6606 22.3306 23.6606 22.7406 23.2506C23.1506 22.8406 23.1506 22.1706 22.7406 21.7606L18.5006 17.5006ZM12.5006 17.5006C10.0106 17.5006 8.00063 15.4906 8.00063 13.0006C8.00063 10.5106 10.0106 8.50063 12.5006 8.50063C14.9906 8.50063 17.0006 10.5106 17.0006 13.0006C17.0006 15.4906 14.9906 17.5006 12.5006 17.5006Z"
            fill="#C3CAD9"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <input
        class="chakra-input css-1m48l39"
        id="field-:r2:"
        name="search"
        placeholder="Search Tasks"
        type="text"
      />
      <div
        class="chakra-input__right-element css-2h3fcg"
      >
        <svg
          fill="none"
          height="31"
          viewBox="0 0 30 31"
          width="30"
        >
          <path
            clip-rule="evenodd"
            d="M12.2925 18.9625L16.1725 15.0825L12.2925 11.2025C12.1052 11.0157 12 10.762 12 10.4975C12 10.233 12.1052 9.97933 12.2925 9.7925C12.6825 9.4025 13.3125 9.4025 13.7025 9.7925L18.2925 14.3825C18.6825 14.7725 18.6825 15.4025 18.2925 15.7925L13.7025 20.3825C13.3125 20.7725 12.6825 20.7725 12.2925 20.3825C11.9125 19.9925 11.9025 19.3525 12.2925 18.9625Z"
            fill="#C3CAD9"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
